### PR TITLE
Fixes all clippy and test errors and warnings

### PR DESCRIPTION
**Pull request by fhaynes**
A few things that will be used in the future I prefaced with a _. Variables (mainly results we were throwing away) I prefaces with _result so that we can go through and handle them properly. This also installs clippy in travis and runs it. Any warnings result in a failed build.

### CODE REVIEW

File: src/infrastructure/fragmenter.rs [hunk 1]
```diff
@@ -125,7 +125,7 @@ impl Fragmentation {
             cursor.read_to_end(&mut payload)?;
 
             // add the payload from the fragment to the buffer whe have in cache
-            reassembly_data.buffer.write(payload.as_slice())?;
+            let _bytes_written = reassembly_data.buffer.write(payload.as_slice())?;
```
Comment (TimonPost) [inline src/infrastructure/fragmenter.rs]: this is not used. You could use `write_all` here if you don't use return type.

File: src/infrastructure/channels/reliable_channel.rs [hunk 2]
```diff
@@ -35,7 +35,7 @@ pub struct ReliableChannel {
     // congestion control
     rtt_measurer: RttMeasurer,
     congestion_data: SequenceBuffer<CongestionData>,
-    quality: NetworkQuality,
+    _quality: NetworkQuality,
```
Comment (LucioFranco) [inline src/infrastructure/channels/reliable_channel.rs]: Do we have an issue open for this, if so can we add a TODO here?

File: src/infrastructure/channels/reliable_channel.rs [hunk 3]
```diff
@@ -141,8 +145,8 @@ impl Channel for ReliableChannel {
 
         self.dropped_packets = dropped_packets.into_iter().map(|(_, p)| p).collect();
 
-        let a = buffer.len();
-        let b = acked_header.size();
+        let _ = buffer.len();
```
Comment (LucioFranco) [inline src/infrastructure/channels/reliable_channel.rs]: if we don't need these can we remove them? If we have them for a reason can we add a comment.

File: src/net/connection/connection_pool.rs [hunk 4]
```diff
@@ -61,6 +57,7 @@ impl ConnectionPool {
     }
 
     // Get the number of connected clients.
+    #[allow(dead_code)]
```
Comment (LucioFranco) [inline src/net/connection/connection_pool.rs]: Don't understand why this is dead code? if it's not exposed publicly or used internally we should remove it.

File: src/net/link_conditioner.rs [hunk 5]
```diff
@@ -21,17 +23,19 @@ impl LinkConditioner {
...
+
     /// Function that checks to see if a packet should be dropped or not
     pub fn should_send(&self) -> bool {
-        let r = random::<f64>();
-        if r <= self.packet_loss {
-            false
-        } else {
-            true
-        }
+        random::<f64>() >= self.packet_loss
```
Comment (LucioFranco) [inline src/net/link_conditioner.rs]: beautiful

File: src/net/udp.rs [hunk 6]
```diff
@@ -15,9 +15,9 @@ use std::error::Error;
 pub struct UdpSocket {
     socket: net::UdpSocket,
     recv_buffer: Vec<u8>,
-    config: Arc<NetworkConfig>,
+    _config: Arc<NetworkConfig>,
```
Comment (LucioFranco) [inline src/net/udp.rs]: Can we add a TODO here?

File: src/packet/header/header_parser.rs [hunk 7]
```diff
@@ -1,6 +1,6 @@
 pub trait HeaderParser {
     type Output;
 
-    /// Write the header to the given buffer.
-    fn parse(&self, mut buffer: &mut Vec<u8>) -> Self::Output;
+    // Write the header to the given buffer.
```
Comment (LucioFranco) [inline src/packet/header/header_parser.rs]: Could you explain why you removed the doccomment?

File: src/packet/packet_data.rs [hunk 8]
```diff
@@ -17,8 +17,8 @@ impl PacketData {
     /// Add fragment to this packet
     pub fn add_fragment(&mut self, fragment: &[u8], fragment_data: &[u8]) -> NetworkResult<()> {
         let mut part = Vec::with_capacity(fragment.len() + fragment_data.len());
-        part.write(fragment)?;
-        part.write(fragment_data)?;
+        let _bytes_written = part.write(fragment)?;
```
Comment (LucioFranco) [inline src/packet/packet_data.rs]: 👍 for `Write::write_all` here

File: src/packet/header/header_parser.rs (hunk 7 above)
Comment (LucioFranco) [reply src/packet/header/header_parser.rs]: strange, can we disable that clippy warning?

here is an example of docstrings on a trait https://github.com/tower-rs/tower/blob/master/tower-service/src/lib.rs#L201

File: src/infrastructure/channels/reliable_channel.rs (hunk 2 above)
Comment (fhaynes) [reply src/infrastructure/channels/reliable_channel.rs]: #96

File: src/net/udp.rs (hunk 6 above)
Comment (fhaynes) [reply src/net/udp.rs]: #97

File: examples/simple_udp.rs [hunk 9]
```diff
@@ -63,7 +63,7 @@ pub fn main() {
     // Moving to lat: 10.555, long: 10.55454, alt: 1.3
     // Moving to lat: 5.4545, long: 3.344, alt: 1.33
     // Received text: "Some information"
-    handle.join();
+    let _ = handle.join();
```
Comment (TimonPost) [inline examples/simple_udp.rs]: this one

File: examples/simple_udp.rs [hunk 10]
```diff
@@ -92,7 +92,7 @@ impl Server {
         let mut udp_socket: UdpSocket = UdpSocket::bind(server_address(), config).unwrap();
 
         // next we could specify if or socket should block the current thread when receiving data or not (default = false)
-        udp_socket.set_nonblocking(false);
+        let _ = udp_socket.set_nonblocking(false);
```
Comment (TimonPost) [inline examples/simple_udp.rs]: this one

File: examples/simple_udp.rs [hunk 11]
```diff
@@ -154,7 +154,7 @@ impl Client {
         let mut udp_socket = UdpSocket::bind(client_address(), config).unwrap();
 
         // next we could specify if or socket should block the current thread when receiving data or not (default = false)
-        udp_socket.set_nonblocking(false);
+        let _ = udp_socket.set_nonblocking(false);
```
Comment (TimonPost) [inline examples/simple_udp.rs]: this one

File: examples/simple_udp.rs [hunk 12]
```diff
@@ -164,8 +164,8 @@ impl Client {
 
         match serialized {
             Ok(raw_data) => {
-                self.udp_socket
-                    .send(Packet::reliable_unordered(server_address(), raw_data));
+                let _result = self.udp_socket
```
Comment (TimonPost) [inline examples/simple_udp.rs]: this one

File: src/net/external_ack.rs [hunk 13]
```diff
@@ -8,11 +8,13 @@
 pub struct ExternalAcks {
     /// the last sequence number we have received from the other side.
     pub last_seq: u16,
+    ///
```
Comment (LucioFranco) [inline src/net/external_ack.rs]: empty comment

**Comment by fhaynes** [comment]
We'll have to live with 3 commits in the master history from this PR; this has already taken way more time than it merits.

**Pull request merged**